File: sdk/php/src/Room.php
<?php
// Plantquest PHP Room
namespace PlantquestSDK;

class Room {
    private $sdk;
    public $data = [];
    public $def = ['name' => 'room'];

    public function __construct($sdk, $data = []) {
        $this->sdk = $sdk;
        $this->data = $data;
    }

    private function handleResult($op, $response, $handler) {
        $status = $response['status'];

        if ($status == 200) {
            $json = $response['json'];
            return $handler($json);
        } else {
            throw new Exception('HTTP-ERROR: ' . $op . ': room: ' . $status);
        }
    }

    public function save($data) {
        $op = 'save';
        $this->data = $data;

        $spec = $this->sdk->fetchSpec($op, $this);
        $response = $this->sdk->options['fetch']($spec['url'], $spec);

        return $this->handleResult($op, $response, function($json) {
            $this->data = $json;
            return $this;
        });
    }

    public function load($data) {
        $op = 'load';
        $this->data = $data;

        $spec = $this->sdk->fetchSpec($op, $this);
        $response = $this->sdk->options['fetch']($spec['url'], $spec);

        return $this->handleResult($op, $response, function($json) {
            $this->data = $json;
            return $this;
        });
    }
}

?>

File: sdk/ruby/lib/plantquest_sdk/room.rb

# Plantquest Ruby Room

require 'json'
require 'net/http'

class Room
  attr_accessor :sdk, :def, :data

  def initialize(sdk, def_hash)
    @sdk = sdk
    @def = def_hash
  end

  def save(data)
    self.data = data
    # TODO: validate data

    body = sdk.body('load', self)
    method = sdk.method('load', self)
    url = sdk.endpoint('load', self)
    
    response = sdk.options[:fetch].call(url, method, sdk.options[:apikey], body)
    status = response.code.to_i

    if status == 200
      json = JSON.parse(response.body, symbolize_names: true)
      # TODO: handle errors
      self.data = json[:ent]
      return self
    else
      raise "HTTP-ERROR: save: Room: #{status}"
    end
  end


  def load(data)
    self.data = data
    # TODO: check if data.id is defined

    body = sdk.body('load', self)
    method = sdk.method('load', self)
    url = sdk.endpoint('load', self)

    response = sdk.options[:fetch].call(url, method, sdk.options[:apikey], body)
    status = response.code.to_i

    if status == 200
      json = JSON.parse(response.body, symbolize_names: true)
      # TODO: handle errors
      self.data = json
    else
      raise "HTTP-ERROR: load: Room: #{status}"
    end

    self
  end

  def list(data = {})
    self.data = data
    # TODO: check if data.id is defined

    body = sdk.body('load', self)
    method = sdk.method('load', self)
    url = sdk.endpoint('load', self)

    response = sdk.options[:fetch].call(url, method, sdk.options[:apikey], body)
    status = response.code.to_i

    if status == 200
      json = JSON.parse(response.body, symbolize_names: true)
      # TODO: handle errors
      self.data = json
    else
      raise "HTTP-ERROR: load: Room: #{status}"
    end

    self
  end

end

